[gui][addons] show yes/no dialog about enable of broken addon and improve GUI

AlwinEsch opened the pull request:
## Description

Commit 1:
----------------------------------------

**[addons] replace addon.xml `<broken>` with `<lifecyclestate ...>`**

This is used to take advantage of several different types of status.
With `<broken>` before, only a yes or no (if this is not available) could be given.

The counterpart to `<broken>Some Text</broken>` would now be:
- `<lifecyclestate type="broken">Some Text</lifecyclestate>`

In addition, this now also allows:
- `<lifecyclestate type="deprecated">Some Text</lifecyclestate>`
  - To note addon complete usable, but deprecated and maybe replaced by another one
- `<lifecyclestate type="normal">Some Text</lifecyclestate>`
  - To show as normal, this not really needed as an addon without always "normal"
  - Maybe good to use, if the normal addon has a special condition where good to show in GUI

Another change to the previous one is that specified texts can now also be entered in different languages using e.g. `lang="en_GB"`.

Commit 2:
----------------------------------------

**[addons][gui] add dialog Yes/No for ask use of deprecated/broken addon**

This adds a display of a Yes/No dialog which is shown when the user wants to activate an outdated or broken addon using the GUI (context menu / addon dialog).

He is then asked whether it should really be activated and is shown add-on text which describes why it is out of date or broken.

This function is added in a new file / class because it is used in several places (currently 2)
and may use other places in the future, and similar things can be added to the new class to keep a better overview.

Commit 3:
----------------------------------------
**[addons] show notification after install if deprecated addon**

This shows a notification dialog if an add-on is installed by a user or an automatic process which has been declared as deprecated.

This means that the user receives this information at least once via the GUI and log entry that something has changed.

Commit 4:
----------------------------------------
**[addons][filesystem] add set as deprecated to "Addon.Status" on file item property**

This add the set of "Addon.Status" also if deprecated and add the string "Deprecated" to this value.

Commit 5:
----------------------------------------
**[guiinfo] add new "ListItem.AddonLifecycleType" and "ListItem.AddonLifecycleDesc"**

This adds two new values to the Kodi GUI which can then be queried in the skin.

1. ListItem.AddonLifecycleType
  - This returns the life cycle of the addon as translated strings.
  - These values can be:
    - "Normal" by string id 24169
    - "Deprecated" by string id 24170
    - "Broken" by string id 24171
  - Example how used on skin:
    - `String.IsEqual(ListItem.AddonLifecycleType,$LOCALIZE[24170])`

2. ListItem.AddonLifecycleDesc
  - This returns the text which is described in addon with regard to its lifecycle and to allow its display in the skin.
    - On addon.xml is it e.g. text within `<lifecyclestate type="deprecated">Deprecated and replaced by audiodecoder.sacd addon!</lifecyclestate>`

Commit 6:
----------------------------------------

**[skin][estuary] allow show of deprecated/broken within addon list and dialog**

About dialog was before only possible about broken, with change now also the
deprecated can be shown over addon icon.

With addon list is the new support added to show broken or deprecated over
addon image (like before on addon dialog).

Further are in addon list the enable icons improved where it shows different
one for different lifecycles of addon.
- Normal hook if nothing special against
- Hook with exclamation mark if addon deprecated
- Broken hook if addon marked as broken

> Note: This icons are only shown if enabled, by disabled still the normal cross shown.

Commit 7:
----------------------------------------

**[jsonrpc] add support to get addon deprecated status**

This add the new value "deprecated" to "Addon.Fields" and "Addon.Details".

Used as new value and on end to prevent compatibility problems and allow further use
of the "broken" value there.

Thought to allow a outside app/site part also to show this.

Commit 8:
----------------------------------------

**add lifecyclestate to xbmc.addon/metadata.xsd**

This to allow check about new addon.xml's






<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## User related
- Improvement: Asks you if you have an add-on marked as broken in your system and want to activate it
- Improvement: New GUI list items added

  | Name | Description
  |------|------------
  | ListItem.AddonLifecycleType | To get related type string where also can be "broken", "deprecated" and "normal"
  | ListItem.AddonLifecycleDesc | To get the text where was before on "ListItem.AddonBroken"
  > *NOTE:* `ListItem.AddonBroken` deprecated, but still supported
  - New json webinterface **shema** type for `deprecated`within `Addon.Details` and `Addon.Fields`
- Improvement (skin.estuary): Shows broken or deprecated declared in addon list using addon picture and list icon


## Motivation and Context
Found during change of vfs.sacd to audiodecoder.sacd, that if installed vfs is marked as broken a user can't direct see (only in addon image on dialog) and can be overlooked

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

### Yes/No dialog show if selected via context menu
![Bildschirmfoto von 2020-08-12 14-25-53](https://user-images.githubusercontent.com/6879739/90055686-07353780-dcde-11ea-8297-75558703749c.png)

### Yes/No dialog show if selected via addon dialog
![Bildschirmfoto von 2020-08-12 14-26-04](https://user-images.githubusercontent.com/6879739/90055689-08fefb00-dcde-11ea-8a38-41ede2d78d95.png)

### New list view if broken on disabled (like them in list was also before about enabled)
![Bildschirmfoto von 2020-08-12 20-40-36](https://user-images.githubusercontent.com/6879739/90055712-13b99000-dcde-11ea-8230-469282943b33.png)

### New list view if broken on enabled

![Bildschirmfoto von 2020-08-12 20-40-25](https://user-images.githubusercontent.com/6879739/90055695-0b615500-dcde-11ea-91da-22332acb68c8.png)

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

garbear commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -15727,7 +15727,19 @@ msgctxt "#24163"
+#. Yes/no dialog text with addon broken string value and to ask for use
+#: xbmc/addons/GUIDialogAddonInfo.cpp
+msgctxt "#24165"
+msgid "Addon marked as broken with following note:[CR][B][I][LIGHT]{0:s}[/LIGHT][/I][/B][CR][CR]Do you want to enable?"
```
The drastic font change is a little jarring. I would just do italics and drop the bold. Italics results in a thinner, lighter font anyway, so `[LIGHT]` might not be required

AlwinEsch replied:
I thought I'd really like to highlight the text explicitly so that the user can quickly see what's wrong with it.
Using an add-on declared as broken can be dangerous.

Has rechecked now with all available skins, the `[LIGHT]` is really not needed and removed. The other two are seems good to have.

With **[I]{0:s}[/I]** only:
![Bildschirmfoto von 2020-08-13 04-46-34](https://user-images.githubusercontent.com/6879739/90088959-80ee1500-dd20-11ea-903a-b6e55cf6a8c6.png)

With **[LIGHT]{0:s}[/LIGHT]** only:
![Bildschirmfoto von 2020-08-13 04-48-10](https://user-images.githubusercontent.com/6879739/90088963-82b7d880-dd20-11ea-849c-ca14624d25b9.png)

With **[B]{0:s}[/B]** only:
![Bildschirmfoto von 2020-08-13 04-49-08](https://user-images.githubusercontent.com/6879739/90088969-864b5f80-dd20-11ea-80e9-62d15edef9be.png)

AlwinEsch replied:
Funny was also on all other skins can't seen that a broken one is enabled, only was there in addon dialog the button "Disable" present :rofl:

ronie reviewed (approved):
nice, skin changes look good to me. thx!

ksooo commented:
To my understanding, if an add-on is broken, it can't be installed for technical reasons? In your case with the VFS add-on, why is it marked broken. Are you "misusing" broken for deprecation, which are imo different things, semantically.

Broken means does not work anymore, bet it for technical or functional reasons.

Deprecated means does still work, but will stop working on the future.

phunkyfish commented:
It almost feels like there should be a state here. Like DEFAULT, BROKEN, DEPRECATED. Similar to disabled reason.

If the addon is deprecated is it disabled or does it still function?

AlwinEsch commented:
Has not thought so far. Seen only here https://kodi.wiki/view/Addon.xml as `<broken>deprecated</broken>` :smile: 

But is right with them inside it it also no more availlable and possible to install from mirror. Think then to add support about `<deprecated>...</deprecated>` on addon.xml.

Or maybe to start kickout of `<broken>` (with still available backward support) and to use:
- `<addonstate type="broken">SOME TEXT</addonstate>`
- `<addonstate type="deprecated">SOME TEXT</addonstate>`
- `<addonstate type="replaced">SOME TEXT</addonstate>`
- `<addonstate type="deprecated;replaced">SOME TEXT</addonstate>` (To mark maybe also two possible fields)
- `<addonstate type="test">SOME TEXT</addonstate>`

Think thats better as his "type" text maybe also usable on GUI.

phunkyfish commented:
I would probably use `replacedWith` instead of `replacer`.

But I'm not sure replaced is a good field to have. It means that one add-on can override another.

ksooo commented:
> It means that one add-on can override another.

Which is imo a nogo.

phunkyfish commented:
Agreed

AlwinEsch commented:
Is updated and also use now deprecated to addon. See changed requests docs above.
With them a change in GUI List items (old stil supported) and change in json part.

Here how look now:
![Bildschirmfoto von 2020-08-15 19-36-23](https://user-images.githubusercontent.com/6879739/90318976-7e640900-df34-11ea-9a13-394d321a282f.png)

ronie commented on addons/skin.estuary/xml/AddonBrowser.xml:
```diff
@@ -20,6 +20,38 @@
+						<height>470</height>
+						<align>center</align>
+						<aligny>center</aligny>
+						<visible>String.IsEqual(ListItem.AddonLifecycleType,deprecated)</visible>
```
same here

ronie commented on addons/skin.estuary/xml/AddonBrowser.xml:
```diff
@@ -20,6 +20,38 @@
+						<height>470</height>
+						<align>center</align>
+						<aligny>center</aligny>
+						<visible>String.IsEqual(ListItem.AddonLifecycleType,broken)</visible>
```
and here

ronie commented on addons/skin.estuary/xml/DialogAddonInfo.xml:
```diff
@@ -213,7 +213,7 @@
 				</control>
 				<control type="group">
-					<visible>!String.IsEmpty(ListItem.AddonBroken)</visible>
+					<visible>!String.IsEqual(ListItem.AddonLifecycleType,normal)</visible>
```
here as well

garbear commented:
> ![Bildschirmfoto von 2020-08-15 19-36-23](https://user-images.githubusercontent.com/6879739/90318976-7e640900-df34-11ea-9a13-394d321a282f.png)

I like the cute broken icon

ronie commented on addons/skin.estuary/xml/DialogAddonInfo.xml:
```diff
@@ -228,6 +228,18 @@
 						<height>500</height>
 						<align>center</align>
 						<aligny>center</aligny>
+						<visible>String.IsEqual(ListItem.AddonLifecycleType,deprecated)</visible>
```
same

ronie commented on addons/skin.estuary/xml/DialogAddonInfo.xml:
```diff
@@ -228,6 +228,18 @@
+						<height>500</height>
+						<align>center</align>
+						<aligny>center</aligny>
+						<visible>String.IsEqual(ListItem.AddonLifecycleType,broken)</visible>
```
and again

ronie commented on addons/skin.estuary/xml/Variables.xml:
```diff
@@ -160,7 +160,8 @@
 	</variable>
 	<variable name="AddonsListIconVar">
-		<value condition="!String.IsEmpty(ListItem.AddonBroken)">icons/addonstatus/disable.png</value>
+		<value condition="(String.IsEqual(ListItem.AddonLifecycleType,deprecated) | String.IsEqual(ListItem.AddonLifecycleType,broken))+ !ListItem.Property(addon.isenabled)">icons/addonstatus/disable.png</value>
```
and here too

ronie commented on addons/skin.estuary/xml/Variables.xml:
```diff
@@ -160,7 +160,8 @@
 	<variable name="AddonsListIconVar">
-		<value condition="!String.IsEmpty(ListItem.AddonBroken)">icons/addonstatus/disable.png</value>
+		<value condition="(String.IsEqual(ListItem.AddonLifecycleType,deprecated) | String.IsEqual(ListItem.AddonLifecycleType,broken))+ !ListItem.Property(addon.isenabled)">icons/addonstatus/disable.png</value>
+		<value condition="ListItem.Property(addon.isenabled)">$INFO[ListItem.AddonLifecycleType,icons/addonstatus/enabled-,.png]</value>
```
sorry, this is not going to works as $INFO[ListItem.AddonLifecycleType] needs to return a localized string

AlwinEsch replied:
Thought to reduce the work amount and where I asked you on slack to use this identifier to get related image about.
Also not seen them as usable text, only as type identifiers.

Can then use as;
```xml
<value condition="ListItem.Property(addon.isenabled)" + String.IsEqual(ListItem.AddonLifecycleType,$LOCALIZE[xxxx1])">icons/addonstatus/enabled-normal.png</value>
<value condition="ListItem.Property(addon.isenabled)" + String.IsEqual(ListItem.AddonLifecycleType,$LOCALIZE[xxxx2])">icons/addonstatus/enabled-deprecated.png</value>
<value condition="ListItem.Property(addon.isenabled)" + String.IsEqual(ListItem.AddonLifecycleType,$LOCALIZE[xxxx3])">icons/addonstatus/enabled-broken.png</value>
```
Can be this much string compares then everywhere not touch the performance?

-----------------------------

Thought before for use e.g. like this
Also thought for this cases that addon can use his own labels, e.g.
```xml
<control type="label">
  <left>85</left>
  <right>40</right>
  <bottom>0</bottom>
  <aligny>center</aligny>
  <visible>String.IsEqual(ListItem.AddonLifecycleType,normal)</visible>
  <label>31001</label>
</control>
<control type="label">
  <left>85</left>
  <right>40</right>
  <bottom>0</bottom>
  <aligny>center</aligny>
  <visible>String.IsEqual(ListItem.AddonLifecycleType,deprecated)</visible>
  <label>31002</label>
</control>
<control type="label">
  <left>85</left>
  <right>40</right>
  <bottom>0</bottom>
  <aligny>center</aligny>
  <visible>String.IsEqual(ListItem.AddonLifecycleType,broken)</visible>
  <label>31003</label>
</control>
```

ronie replied:
instead of providing ListItem.AddonLifecycleType as an infolabel and then using String.IsEqual() to convert it to a boolean,
why not provide an infobool directly?
something like `ListItem.HasAddonLifecycleType(deprecated)`

ronie commented on addons/skin.estuary/xml/AddonBrowser.xml:
```diff
@@ -20,6 +20,38 @@
+						<label>$LOCALIZE[24049]</label>
+						<font>font36_title</font>
+					</control>
+					<control type="textbox">
```
instead of duplicating the textbox, you can remove the visible condition and use variable for the label.
`<label>$VAR[AddonLifecycleType]</label>`

and in Variables.xml:
```
<variable name="AddonLifecycleType">
	<value condition="String.IsEqual(ListItem.AddonLifecycleType,$LOCALIZE[xxx])">$LOCALIZE[24049]</value>
	<value condition="String.IsEqual(ListItem.AddonLifecycleType,$LOCALIZE[yyy])">$LOCALIZE[24096]</value>
</variable>
```

AlwinEsch replied:
Yeah, that makes sense :+1:

ronie commented on addons/skin.estuary/xml/DialogAddonInfo.xml:
```diff
@@ -228,6 +228,18 @@
+						<label>$LOCALIZE[24049]</label>
+						<font>font36_title</font>
+					</control>
+					<control type="textbox">
```
no need to duplicate the textbox here either.

AlwinEsch commented:
Becomes changed tomorrow :smile: 

Only confuse me that in GUIInfoManager.cpp mostly nowhere a part with string id's is noted :smirk: 

Found only this but that the opposite way :thinking:  :
```cpp
///   \table_row3{   <b>`Playlist.Repeat`</b>,
///                  \anchor Playlist_Repeat
///                  _integer_,
///     @return The String Id for the repeat mode. It can be one of the following
///     values:
///       - **592** (Repeat One)
///       - **593** (Repeat All)
///       - **594** (Repeat Off)
///     <p>
///   }
```

phunkyfish commented on xbmc/addons/ContextMenus.cpp:
```diff
@@ -62,6 +63,33 @@ bool CEnableAddon::IsVisible(const CFileItem& item) const
 
 bool CEnableAddon::Execute(const CFileItemPtr& item) const
 {
+  int header_nr;
+  int text_nr;
+  switch (item->GetAddonInfo()->LifecycleState())
+  {
+    case AddonLifecycleState::BROKEN:
+      header_nr = 24164;
+      text_nr = 24165;
+      break;
+    case AddonLifecycleState::DEPRECATED:
+      header_nr = 24166;
+      text_nr = 24167;
+      break;
+    default:
+      header_nr = 0;
+      text_nr = 0;
+      break;
+  }
+  if (header_nr > 0)
+  {
+    std::string header =
+        StringUtils::Format(g_localizeStrings.Get(header_nr), item->GetAddonInfo()->ID());
+    std::string text = StringUtils::Format(g_localizeStrings.Get(text_nr),
+                                           item->GetAddonInfo()->LifecycleStateDescription());
+    if (!CGUIDialogYesNo::ShowAndGetInput(header, text))
+      return false;
+  }
```
Duplicates the code in `GUIDialogAddonInfo.cpp`. Does it have to?

phunkyfish commented on xbmc/addons/GUIDialogAddonInfo.cpp:
```diff
@@ -454,7 +455,36 @@ void CGUIDialogAddonInfo::OnEnableDisable()
   }
   else
+  {
+    int header_nr;
+    int text_nr;
+    switch (m_localAddon->LifecycleState())
+    {
+      case AddonLifecycleState::BROKEN:
+        header_nr = 24164;
+        text_nr = 24165;
+        break;
+      case AddonLifecycleState::DEPRECATED:
+        header_nr = 24166;
+        text_nr = 24167;
+        break;
+      default:
+        header_nr = 0;
+        text_nr = 0;
+        break;
+    }
+    if (header_nr > 0)
+    {
+      std::string header =
+          StringUtils::Format(g_localizeStrings.Get(header_nr), m_localAddon->ID());
+      std::string text = StringUtils::Format(g_localizeStrings.Get(text_nr),
+                                             m_localAddon->LifecycleStateDescription());
+      if (!CGUIDialogYesNo::ShowAndGetInput(header, text))
+        return;
+    }
```
Duplicates the code in `ContextMenus.cpp`. Does it have to?

phunkyfish commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -47,6 +47,23 @@ enum class AddonOriginType
 };
 
+/*!
+ * @brief Add-on state defined within addon.xml to report about special cases
```
```
/*!
 * Add-on state defined within addon.xml to report about the current addon lifecycle state
 *
 * E.g. the add-on is broken and can no longer be used.
 *
```

phunkyfish commented on xbmc/addons/addoninfo/AddonInfo.h:
```diff
@@ -47,6 +47,23 @@ enum class AddonOriginType
+ */
+enum class AddonLifecycleState
+{
+  NORMAL = 0, /// Used if add-on without any cases and in normal condition (no field within addon)
+  BROKEN = 1, /// Mark the add-on as broken in the Kodi repo
+  DEPRECATED = 2, /// To set the add-on as deprected and still usable (maybe replaced with another)
```
```
  NORMAL = 0, /// Used if an add-on has no special lifecycle state, i.e. 
  BROKEN = 1, /// the add-on should marked as broken in the repository
  DEPRECATED = 2, /// the add-on should be marked as deprecated but is still usable
```

DaveTBlake commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -1618,7 +1618,7 @@
     "items": { "type": "string",
       "enum": [ "name", "version", "summary", "description", "path", "author", "thumbnail", "disclaimer", "fanart",
-                "dependencies", "broken", "extrainfo", "rating", "enabled", "installed" ]
+                "dependencies", "broken", "deprecated", "extrainfo", "rating", "enabled", "installed" ]
```
Fields need to be added at the end of enum list, and not inserted

DaveTBlake commented on xbmc/interfaces/json-rpc/schema/version.txt:
```diff
@@ -1 +1 @@
-JSONRPC_VERSION 11.12.1
+JSONRPC_VERSION 11.12.2
```
As a backward compatible schema change this need to bump minor not patch e.g. 11.13.0

https://github.com/xbmc/xbmc/blob/04ba4e0fd6d87b5a0fec68b22d96a1049a33128f/xbmc/interfaces/json-rpc/schema/methods.json#L32-L35

DaveTBlake commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -1645,6 +1645,7 @@
         }
       },
       "broken": { "type": [ "boolean", "string" ] },
+      "deprecated": { "type": [ "boolean", "string" ] },
```
At end too

DaveTBlake reviewed (commented):
Some comments on the JSON API changes

AlwinEsch committed:
[addons] replace addon.xml <broken> with <lifecyclestate ...>

This is used to take advantage of several different types of status.
With `<broken>` before, only a yes or no (if this is not available) could be given.

The counterpart to `<broken>Some Text</broken>` would now be:
- `<lifecyclestate type="broken">Some Text</lifecyclestate>`

In addition, this now also allows:
- `<lifecyclestate type="deprecated">Some Text</lifecyclestate>`
  - To note addon complete usable, but deprecated and maybe replaced by another one
- `<lifecyclestate type="normal">Some Text</lifecyclestate>`
  - To show as normal, this not really needed as an addon without always "normal"
  - Maybe good to use, if the normal addon has a special condition where good to show in GUI

Another change to the previous one is that specified texts can now also be entered in different languages using e.g. `lang="en_GB"`.

AlwinEsch committed:
[addons] show notification after install if deprecated addon

This shows a notification dialog if an add-on is installed by a user or an automatic process which has been declared as deprecated.

This means that the user receives this information at least once via the GUI and log entry that something has changed.

AlwinEsch committed:
[addons][filesystem] add set as deprecated to "Addon.Status" on file item property

This add the set of "Addon.Status" also if deprecated and add the string "Deprecated" to this value.

AlwinEsch committed:
[guiinfo] add new "ListItem.AddonLifecycleType" and "ListItem.AddonLifecycleDesc"

This adds two new values to the Kodi GUI which can then be queried in the skin.

1. ListItem.AddonLifecycleType
  - This returns the life cycle of the addon as translated strings.
  - These values can be:
    - "Normal" by string id 24169
    - "Deprecated" by string id 24170
    - "Broken" by string id 24171
  - Example how used on skin:
    - `String.IsEqual(ListItem.AddonLifecycleType,$LOCALIZE[24170])`

2. ListItem.AddonLifecycleDesc
  - This returns the text which is described in addon with regard to its lifecycle and to allow its display in the skin.
    - On addon.xml is it e.g. text within `<lifecyclestate type="deprecated">Deprecated and replaced by audiodecoder.sacd addon!</lifecyclestate>`

AlwinEsch committed:
[jsonrpc] add support to get addon deprecated status

This add the new value "deprecated" to "Addon.Fields" and "Addon.Details".

Used as new value and on end to prevent compatibility problems and allow further use
of the "broken" value there.

Thought to allow a outside app/site part also to show this.

ronie commented on addons/skin.estuary/xml/Variables.xml:
```diff
@@ -159,13 +159,19 @@
+	</variable>
 	<variable name="AddonsListIconVar">
-		<value condition="!String.IsEmpty(ListItem.AddonBroken)">icons/addonstatus/disable.png</value>
+		<value condition="(String.IsEqual(ListItem.AddonLifecycleType,$LOCALIZE[24170]) | String.IsEqual(ListItem.AddonLifecycleType,$LOCALIZE[24171])) + !ListItem.Property(addon.isenabled)">icons/addonstatus/disable.png</value>
```
you need to use square brackets around a group of conditions: `[A | B] + !C`

AlwinEsch replied:
Thanks have changed now to them.

AlwinEsch commented on xbmc/addons/AddonDatabase.cpp:
```diff
@@ -81,7 +82,12 @@ static void DeserializeMetadata(const std::string& document, CAddonInfoBuilder::
   builder.SetAuthor(variant["author"].asString());
   builder.SetDisclaimer(variant["disclaimer"].asString());
-  builder.SetBroken(variant["broken"].asString());
+  if (variant.isMember("broken")) // Fallback of old
```
About `addon-check` I created here https://github.com/xbmc/addon-check/pull/240 a request about.

ronie reviewed (approved):
skin changes look good to me

AlwinEsch committed:
[skin][estuary] allow show of deprecated/broken within addon list and dialog

About dialog was before only possible about broken, with change now also the
deprecated can be shown over addon icon.

With addon list is the new support added to show broken or deprecated over
addon image (like before on addon dialog).

Further are in addon list the enable icons improved where it shows different
one for different lifecycles of addon.
- Normal hook if nothing special against
- Hook with exclamation mark if addon deprecated
- Broken hook if addon marked as broken

> Note: This icons are only shown if enabled, by disabled still the normal cross shown.

AlwinEsch committed:
add lifecyclestate to xbmc.addon/metadata.xsd

This to allow check about new addon.xml's

da-anda commented:
to follow up on the `replacedWith` idea. @phunkyfish @ksooo why is this a bad idea? It ofc should not be auto-updated, but having the info about a potential successor skins can show a button/dialog and ask the user if he would like to install the replacement. It always as to be a manual user action and no auto update ofc.

phunkyfish commented:
It’s similar to the idea of specifying a beta repo for an addon. Sure it can be useful but without signing in place can be very difficult to police/establish trust. We don’t want addons to be able to impact each other. This would open a door after we have just closed so many. I’m not saying we don’t do this, just not yet. We’ll have a bit of a journey to get there.

AlwinEsch commented:
Have it rebased and created request in Drupal, after them in there this request should be ready.

With `replacedWith` thing to leave now request as is. With the other we can look on another request/place. More good to have a better trusted addon check before to prevent wrong insert.

AlwinEsch commented:
About our website is also now a request present to bring support about the new "lifecyclestate" and to get "broken" declaration from it (test yesterday here http://xbmcfoundationdev.prod.acquia-sites.com/addons was OK).

It seems then the request ready to use, only then a Wiki update needed afterwards.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch committed:
[addons][gui] add dialog Yes/No for ask use of deprecated/broken addon

This adds a display of a Yes/No dialog which is shown when the user wants to activate an outdated or broken addon using the GUI (context menu / addon dialog).

He is then asked whether it should really be activated and is shown add-on text which describes why it is out of date or broken.

This function is added in a new file / class because it is used in several places (currently 2)
and may use other places in the future, and similar things can be added to the new class to keep a better overview.

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
Ping @enen92, @DaveTBlake and @Montellese.

Is this request OK to come in?

phunkyfish reviewed (approved):
Looks good to me.

phunkyfish commented:
jenkins build this please

enen92 reviewed (approved):
Lgtm

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
OK to bring in?

phunkyfish commented:
You have 3 approvals. I think you are good.

AlwinEsch commented:
Yeah, if I touch parts outside of addons then I'm always a bit angry :sweat_smile:

howie-f commented:
hmm, after this pr addons are marked as broken in the available updates list which is not true:

![screenshot000](https://user-images.githubusercontent.com/58829855/93142659-bfdc0580-f6e6-11ea-96ce-8915584977c6.png)

the addon can still be installed via repo but after downgrading it shows up broken again.
any idea why this is happening? i verified that nether the repo nor the addon has been changed

AlwinEsch commented:
Thanks for the info, I'll see what's going wrong after work and fix it.
